File: agent/src/main/java/com/smancode/smanagent/util/StackTraceUtils.java
package com.smancode.smanagent.util;

import java.io.PrintWriter;
import java.io.StringWriter;

/**
 * 堆栈跟踪工具类
 * <p>
 * 将异常堆栈转换为单行字符串，减少日志行数
 */
public class StackTraceUtils {

    private static final String STACK_TRACE_SEPARATOR = " <- ";

    /**
     * 将异常堆栈转换为单行字符串
     * <p>
     * 格式：异常消息 <- 类名.方法名(行号) <- 类名.方法名(行号) ...
     *
     * @param throwable 异常对象
     * @return 单行堆栈字符串
     */
    public static String formatStackTrace(Throwable throwable) {
        if (throwable == null) {
            return "";
        }

        StringBuilder sb = new StringBuilder();

        // 异常类型和消息
        String message = throwable.getMessage();
        sb.append(throwable.getClass().getSimpleName());
        if (message != null && !message.isEmpty()) {
            sb.append(": ").append(message);
        }

        // 堆栈跟踪元素
        StackTraceElement[] stackTrace = throwable.getStackTrace();
        if (stackTrace != null && stackTrace.length > 0) {
            sb.append(STACK_TRACE_SEPARATOR);
            for (int i = 0; i < stackTrace.length && i < 10; i++) {
                if (i > 0) {
                    sb.append(STACK_TRACE_SEPARATOR);
                }
                sb.append(stackTrace[i].toString());
            }
            if (stackTrace.length > 10) {
                sb.append(STACK_TRACE_SEPARATOR).append("...");
            }
        }

        // 处理 cause
        Throwable cause = throwable.getCause();
        if (cause != null && cause != throwable) {
            sb.append(" [Caused by: ").append(formatStackTrace(cause)).append("]");
        }

        return sb.toString();
    }

    /**
     * 获取完整的堆栈跟踪字符串（多行）
     *
     * @param throwable 异常对象
     * @return 完整堆栈字符串
     */
    public static String getFullStackTrace(Throwable throwable) {
        if (throwable == null) {
            return "";
        }
        StringWriter sw = new StringWriter();
        PrintWriter pw = new PrintWriter(sw);
        throwable.printStackTrace(pw);
        return sw.toString();
    }
}

File: src/main/kotlin/com/smancode/sman/util/StackTraceUtils.kt
package com.smancode.sman.util

import java.io.PrintWriter
import java.io.StringWriter

/**
 * 堆栈跟踪工具类
 */
object StackTraceUtils {

    /**
     * 格式化异常堆栈跟踪
     *
     * @param throwable 异常
     * @return 格式化后的堆栈跟踪字符串
     */
    fun formatStackTrace(throwable: Throwable?): String {
        if (throwable == null) {
            return ""
        }

        val sw = StringWriter()
        val pw = PrintWriter(sw)
        throwable.printStackTrace(pw)
        return sw.toString()
    }

    /**
     * 获取异常根因
     *
     * @param throwable 异常
     * @return 根因异常
     */
    fun getRootCause(throwable: Throwable): Throwable {
        var cause: Throwable? = throwable
        while (cause?.cause != null) {
            cause = cause.cause
        }
        return cause ?: throwable
    }

    /**
     * 获取简洁的堆栈跟踪（只包含关键信息）
     *
     * @param throwable 异常
     * @param maxFrames 最大帧数
     * @return 简洁的堆栈跟踪
     */
    fun getCompactStackTrace(throwable: Throwable?, maxFrames: Int = 10): String {
        if (throwable == null) {
            return ""
        }

        val sb = StringBuilder()
        sb.append(throwable.javaClass.name)
        if (throwable.message != null) {
            sb.append(": ").append(throwable.message)
        }
        sb.append("\n")

        val stackTrace = throwable.stackTrace
        val framesToPrint = minOf(stackTrace.size, maxFrames)

        for (i in 0 until framesToPrint) {
            sb.append("\tat ").append(stackTrace[i]).append("\n")
        }

        if (stackTrace.size > maxFrames) {
            sb.append("\t... ").append(stackTrace.size - maxFrames).append(" more\n")
        }

        return sb.toString()
    }
}
